File: Source/AlphaTab/Rendering/Glyphs/ScoreSlideLineGlyph.cs
using System;
using AlphaTab.Model;
using AlphaTab.Platform;

namespace AlphaTab.Rendering.Glyphs
{
    internal class ScoreSlideLineGlyph : Glyph
    {
        private readonly SlideOutType _outType;
        private readonly SlideInType _inType;
        private readonly Note _startNote;
        private readonly BeatContainerGlyph _parent;

        public ScoreSlideLineGlyph(SlideInType inType, SlideOutType outType, Note startNote, BeatContainerGlyph parent)
            : base(0, 0)
        {
            _outType = outType;
            _inType = inType;
            _startNote = startNote;
            _parent = parent;
        }

        public override void DoLayout()
        {
            Width = 0;
        }

        public override void Paint(float cx, float cy, ICanvas canvas)
        {
            PaintSlideIn(cx, cy, canvas);
            DrawSlideOut(cx, cy, canvas);
        }

        private void PaintSlideIn(float cx, float cy, ICanvas canvas)
        {
            var startNoteRenderer = (ScoreBarRenderer)Renderer;
            var sizeX = 12 * Scale;
            var offsetX = 1 * Scale;
            float startX;
            float startY;
            float endX;
            float endY;

            switch (_inType)
            {
                case SlideInType.IntoFromBelow:
                    endX = cx + startNoteRenderer.X + startNoteRenderer.GetNoteX(_startNote, false) - offsetX;
                    endY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote, true) +
                           NoteHeadGlyph.NoteHeadHeight / 2;

                    startX = endX - sizeX;
                    startY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote, true) +
                             NoteHeadGlyph.NoteHeadHeight;

                    break;
                case SlideInType.IntoFromAbove:
                    endX = cx + startNoteRenderer.X + startNoteRenderer.GetNoteX(_startNote, false) - offsetX;
                    endY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote, true) +
                           NoteHeadGlyph.NoteHeadHeight / 2;

                    startX = endX - sizeX;
                    startY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote, true);
                    break;
                default:
                    return;
            }

            var accidentalsWidth = GetAccidentalsWidth(startNoteRenderer, _startNote.Beat);
            startX -= accidentalsWidth;
            endX -= accidentalsWidth;

            PaintSlideLine(canvas, false, startX, endX, startY, endY);
        }


        private float GetAccidentalsWidth(ScoreBarRenderer renderer, Beat beat)
        {
            var preNotes = (ScoreBeatPreNotesGlyph)renderer.GetPreNotesGlyphForBeat(beat);
            if (preNotes != null && preNotes.Accidentals != null)
            {
                return preNotes.Accidentals.Width;
            }

            return 0;
        }

        private void DrawSlideOut(float cx, float cy, ICanvas canvas)
        {
            var startNoteRenderer = (ScoreBarRenderer)Renderer;
            var sizeX = 12 * Scale;
            var offsetX = 1 * Scale;
            float startX;
            float startY;
            float endX;
            float endY;
            var waves = false;

            switch (_outType)
            {
                case SlideOutType.Shift:
                case SlideOutType.Legato:
                    startX = cx + startNoteRenderer.X +
                             startNoteRenderer.GetBeatX(_startNote.Beat, BeatXPosition.PostNotes) + offsetX;
                    var isUp = _startNote.SlideTarget.RealValue > _startNote.RealValue;
                    startY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote);
                    var lineOffset = 0.25f * NoteHeadGlyph.NoteHeadHeight * Scale;
                    if (isUp)
                    {
                        startY += lineOffset;
                    }
                    else
                    {
                        startY -= lineOffset;
                    }

                    if (_startNote.SlideTarget != null)
                    {
                        var endNoteRenderer =
                            Renderer.ScoreRenderer.Layout.GetRendererForBar<BarRendererBase>(Renderer.Staff.StaveId,
                                _startNote.SlideTarget.Beat.Voice.Bar);
                        if (endNoteRenderer == null || endNoteRenderer.Staff != startNoteRenderer.Staff)
                        {
                            endX = cx + startNoteRenderer.X + _parent.X;
                            endY = startY;
                        }
                        else
                        {
                            endX = cx + endNoteRenderer.X +
                                   endNoteRenderer.GetBeatX(_startNote.SlideTarget.Beat, BeatXPosition.PreNotes) -
                                   offsetX;
                            endY = cy + endNoteRenderer.Y + endNoteRenderer.GetNoteY(_startNote.SlideTarget);
                            if (isUp)
                            {
                                endY -= lineOffset;
                            }
                            else
                            {
                                endY += lineOffset;
                            }
                        }
                    }
                    else
                    {
                        endX = cx + startNoteRenderer.X + _parent.X;
                        endY = startY;
                    }

                    break;
                case SlideOutType.OutUp:
                    startX = cx + startNoteRenderer.X + startNoteRenderer.GetNoteX(_startNote) + offsetX * 2;
                    startY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote, true) +
                             NoteHeadGlyph.NoteHeadHeight / 2;
                    endX = startX + sizeX;
                    endY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote, true);
                    break;
                case SlideOutType.OutDown:
                    startX = cx + startNoteRenderer.X + startNoteRenderer.GetNoteX(_startNote) + offsetX * 2;
                    startY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote, true) +
                             NoteHeadGlyph.NoteHeadHeight / 2;
                    endX = startX + sizeX;
                    endY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote, true) +
                           NoteHeadGlyph.NoteHeadHeight;
                    break;
                case SlideOutType.PickSlideUp:
                    startX = cx + startNoteRenderer.X + startNoteRenderer.GetNoteX(_startNote) + offsetX * 2;
                    startY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote, true) +
                             NoteHeadGlyph.NoteHeadHeight / 2;
                    endY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote, true) -
                           NoteHeadGlyph.NoteHeadHeight;
                    endX = cx + startNoteRenderer.X + startNoteRenderer.Width;

                    if (_startNote.Beat.NextBeat != null && _startNote.Beat.NextBeat.Voice == _startNote.Beat.Voice)
                    {
                        endX = cx + startNoteRenderer.X +
                               startNoteRenderer.GetBeatX(_startNote.Beat.NextBeat, BeatXPosition.PreNotes);
                    }

                    waves = true;
                    break;
                case SlideOutType.PickSlideDown:
                    startX = cx + startNoteRenderer.X + startNoteRenderer.GetNoteX(_startNote) + offsetX * 2;
                    startY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote, true) -
                             NoteHeadGlyph.NoteHeadHeight / 2;
                    endY = cy + startNoteRenderer.Y + startNoteRenderer.GetNoteY(_startNote, true) +
                           NoteHeadGlyph.NoteHeadHeight;
                    endX = cx + startNoteRenderer.X + startNoteRenderer.Width;

                    if (_startNote.Beat.NextBeat != null && _startNote.Beat.NextBeat.Voice == _startNote.Beat.Voice)
                    {
                        endX = cx + startNoteRenderer.X +
                               startNoteRenderer.GetBeatX(_startNote.Beat.NextBeat, BeatXPosition.PreNotes);
                    }

                    waves = true;
                    break;
                default:
                    return;
            }

            PaintSlideLine(canvas, waves, startX, endX, startY, endY);
        }

        private void PaintSlideLine(ICanvas canvas, bool waves, float startX, float endX, float startY, float endY)
        {
            if (waves)
            {
                var b = endX - startX;
                var a = endY - startY;
                var c = Math.Sqrt(Math.Pow(a, 2) + Math.Pow(b, 2));
                var angle = (float)(Math.Asin(a / c) * (180 / Math.PI));
                canvas.BeginRotate(startX, startY, angle);

                var glyph = new NoteVibratoGlyph(0, 0, VibratoType.Slight);
                glyph.Renderer = Renderer;
                glyph.DoLayout();
                glyph.Width = b;
                glyph.Paint(0, 0, canvas);

                canvas.EndRotate();
            }
            else
            {
                canvas.BeginPath();
                canvas.MoveTo(startX, startY);
                canvas.LineTo(endX, endY);
                canvas.Stroke();
            }
        }
    }
}

File: src/rendering/glyphs/ScoreSlideLineGlyph.ts
import type { Beat } from '@src/model/Beat';
import type { Note } from '@src/model/Note';
import { SlideInType } from '@src/model/SlideInType';
import { SlideOutType } from '@src/model/SlideOutType';
import { VibratoType } from '@src/model/VibratoType';
import type { ICanvas } from '@src/platform/ICanvas';
import { type BarRendererBase, NoteXPosition, NoteYPosition } from '@src/rendering/BarRendererBase';
import { BeatXPosition } from '@src/rendering/BeatXPosition';
import type { BeatContainerGlyph } from '@src/rendering/glyphs/BeatContainerGlyph';
import { Glyph } from '@src/rendering/glyphs/Glyph';
import { NoteVibratoGlyph } from '@src/rendering/glyphs/NoteVibratoGlyph';
import type { ScoreBeatPreNotesGlyph } from '@src/rendering/glyphs/ScoreBeatPreNotesGlyph';
import type { ScoreBarRenderer } from '@src/rendering/ScoreBarRenderer';

export class ScoreSlideLineGlyph extends Glyph {
    private _outType: SlideOutType;
    private _inType: SlideInType;
    private _startNote: Note;
    private _parent: BeatContainerGlyph;

    public constructor(inType: SlideInType, outType: SlideOutType, startNote: Note, parent: BeatContainerGlyph) {
        super(0, 0);
        this._outType = outType;
        this._inType = inType;
        this._startNote = startNote;
        this._parent = parent;
    }

    public override doLayout(): void {
        this.width = 0;
    }

    public override paint(cx: number, cy: number, canvas: ICanvas): void {
        this.paintSlideIn(cx, cy, canvas);
        this.drawSlideOut(cx, cy, canvas);
    }

    private paintSlideIn(cx: number, cy: number, canvas: ICanvas): void {
        const startNoteRenderer: ScoreBarRenderer = this.renderer as ScoreBarRenderer;
        const sizeX: number = 12;
        let endX = cx + startNoteRenderer.x + startNoteRenderer.getNoteX(this._startNote, NoteXPosition.Left) - 2;
        const endY = cy + startNoteRenderer.y + startNoteRenderer.getNoteY(this._startNote, NoteYPosition.Center);
        let startX = endX - sizeX;
        let startY: number = cy + startNoteRenderer.y;

        switch (this._inType) {
            case SlideInType.IntoFromBelow:
                startY += startNoteRenderer.getNoteY(this._startNote, NoteYPosition.Bottom);
                break;
            case SlideInType.IntoFromAbove:
                startY += startNoteRenderer.getNoteY(this._startNote, NoteYPosition.Top);
                break;
            default:
                return;
        }

        const accidentalsWidth: number = this.getAccidentalsWidth(startNoteRenderer, this._startNote.beat);
        startX -= accidentalsWidth;
        endX -= accidentalsWidth;
        this.paintSlideLine(canvas, false, startX, endX, startY, endY);
    }

    private getAccidentalsWidth(renderer: ScoreBarRenderer, beat: Beat): number {
        const preNotes: ScoreBeatPreNotesGlyph = renderer.getPreNotesGlyphForBeat(beat) as ScoreBeatPreNotesGlyph;
        if (preNotes && preNotes.accidentals) {
            return preNotes.accidentals.width;
        }
        return 0;
    }

    private drawSlideOut(cx: number, cy: number, canvas: ICanvas): void {
        const startNoteRenderer: ScoreBarRenderer = this.renderer as ScoreBarRenderer;
        const sizeX: number = 12;
        const offsetX: number = 2;
        const offsetY: number = 2;
        let startX: number = 0;
        let startY: number = 0;
        let endX: number = 0;
        let endY: number = 0;
        let waves: boolean = false;
        switch (this._outType) {
            case SlideOutType.Shift:
            case SlideOutType.Legato:
                startX =
                    cx +
                    startNoteRenderer.x +
                    startNoteRenderer.getBeatX(this._startNote.beat, BeatXPosition.PostNotes) +
                    offsetX;
                startY = cy + startNoteRenderer.y + startNoteRenderer.getNoteY(this._startNote, NoteYPosition.Center);
                if (this._startNote.slideTarget) {
                    const endNoteRenderer: BarRendererBase | null =
                        this.renderer.scoreRenderer.layout!.getRendererForBar(
                            this.renderer.staff.staffId,
                            this._startNote.slideTarget.beat.voice.bar
                        );
                    if (!endNoteRenderer || endNoteRenderer.staff !== startNoteRenderer.staff) {
                        endX = cx + startNoteRenderer.x + startNoteRenderer.width;
                        endY = startY;
                    } else {
                        endX =
                            cx +
                            endNoteRenderer.x +
                            endNoteRenderer.getBeatX(this._startNote.slideTarget.beat, BeatXPosition.PreNotes) -
                            offsetX;
                        endY =
                            cy +
                            endNoteRenderer.y +
                            endNoteRenderer.getNoteY(this._startNote.slideTarget, NoteYPosition.Center);
                    }

                    if (this._startNote.slideTarget.realValue > this._startNote.realValue) {
                        startY += offsetY;
                        endY -= offsetY;
                    } else {
                        startY -= offsetY;
                        endY += offsetY;
                    }
                } else {
                    endX = cx + startNoteRenderer.x + this._parent.x;
                    endY = startY;
                }
                break;
            case SlideOutType.OutUp:
                startX =
                    cx +
                    startNoteRenderer.x +
                    startNoteRenderer.getNoteX(this._startNote, NoteXPosition.Right) +
                    offsetX;
                startY = cy + startNoteRenderer.y + startNoteRenderer.getNoteY(this._startNote, NoteYPosition.Center);
                endX = startX + sizeX;
                endY = cy + startNoteRenderer.y + startNoteRenderer.getNoteY(this._startNote, NoteYPosition.Top);
                break;
            case SlideOutType.OutDown:
                startX =
                    cx +
                    startNoteRenderer.x +
                    startNoteRenderer.getNoteX(this._startNote, NoteXPosition.Right) +
                    offsetX;
                startY = cy + startNoteRenderer.y + startNoteRenderer.getNoteY(this._startNote, NoteYPosition.Center);
                endX = startX + sizeX;
                endY = cy + startNoteRenderer.y + startNoteRenderer.getNoteY(this._startNote, NoteYPosition.Bottom);
                break;
            case SlideOutType.PickSlideUp:
                startX =
                    cx +
                    startNoteRenderer.x +
                    startNoteRenderer.getNoteX(this._startNote, NoteXPosition.Right) +
                    offsetX * 2;
                startY = cy + startNoteRenderer.y + startNoteRenderer.getNoteY(this._startNote, NoteYPosition.Center);
                endY = cy + startNoteRenderer.y + startNoteRenderer.getNoteY(this._startNote, NoteYPosition.Top);
                endX = cx + startNoteRenderer.x + startNoteRenderer.width;
                if (
                    this._startNote.beat.nextBeat &&
                    this._startNote.beat.nextBeat.voice === this._startNote.beat.voice
                ) {
                    endX =
                        cx +
                        startNoteRenderer.x +
                        startNoteRenderer.getBeatX(this._startNote.beat.nextBeat, BeatXPosition.PreNotes);
                }
                waves = true;
                break;
            case SlideOutType.PickSlideDown:
                startX =
                    cx +
                    startNoteRenderer.x +
                    startNoteRenderer.getNoteX(this._startNote, NoteXPosition.Right) +
                    offsetX * 2;
                startY = cy + startNoteRenderer.y + startNoteRenderer.getNoteY(this._startNote, NoteYPosition.Center);
                endY = cy + startNoteRenderer.y + startNoteRenderer.getNoteY(this._startNote, NoteYPosition.Bottom);
                endX = cx + startNoteRenderer.x + startNoteRenderer.width;
                if (
                    this._startNote.beat.nextBeat &&
                    this._startNote.beat.nextBeat.voice === this._startNote.beat.voice
                ) {
                    endX =
                        cx +
                        startNoteRenderer.x +
                        startNoteRenderer.getBeatX(this._startNote.beat.nextBeat, BeatXPosition.PreNotes);
                }
                waves = true;
                break;
            default:
                return;
        }
        this.paintSlideLine(canvas, waves, startX, endX, startY, endY);
    }

    private paintSlideLine(
        canvas: ICanvas,
        waves: boolean,
        startX: number,
        endX: number,
        startY: number,
        endY: number
    ): void {
        if (waves) {
            const glyph: NoteVibratoGlyph = new NoteVibratoGlyph(0, 0, VibratoType.Slight, 1.2);
            glyph.renderer = this.renderer;
            glyph.doLayout();

            startY -= glyph.height / 2;
            endY -= glyph.height / 2;

            const b: number = endX - startX;
            const a: number = endY - startY;
            const c: number = Math.sqrt(Math.pow(a, 2) + Math.pow(b, 2));
            glyph.width = b;

            const angle: number = Math.asin(a / c) * (180 / Math.PI);
            canvas.beginRotate(startX, startY, angle);
            glyph.paint(0, 0, canvas);
            canvas.endRotate();
        } else {
            canvas.beginPath();
            canvas.moveTo(startX, startY);
            canvas.lineTo(endX, endY);
            canvas.stroke();
        }
    }
}
